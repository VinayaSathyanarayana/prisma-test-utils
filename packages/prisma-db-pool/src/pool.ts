import { LiftEngine, DataSource } from '@prisma/lift'
import { DMMF } from '@prisma/photon/runtime/dmmf-types'
import * as fs from 'fs'
import * as os from 'os'
import * as path from 'path'
import pify from 'pify'
import { withDefault } from './utils'

const writeFile = pify(fs.writeFile)
const makeDir = pify(fs.mkdir)

export type PoolDefinition = {
  dmmf: DMMF.Document
  pool: {
    min?: number
    max: number
  }
  cwd?: string
}

export type DBInstance = {
  cwd: string
  datamodel: string
}

export class Pool {
  private dmmf: DMMF.Document
  private cwd: string
  private dbs: {
    booting: string[]
    idle: DBInstance[]
    busy: DBInstance[]
  }
  private waiters: Waiter[]
  private capacity: number

  constructor(definition: PoolDefinition) {
    this.dmmf = definition.dmmf
    this.cwd = withDefault(process.cwd(), definition.cwd)
    this.dbs = {
      booting: [],
      idle: [],
      busy: [],
    }
    this.waiters = []
    this.capacity = definition.pool.max

    /* Start init */

    for (let index = 0; index < withDefault(0, definition.pool.min); index++) {
      this.createDBInstance().then(db => {
        this.dbs.idle = [...this.dbs.idle, db]
      })
    }
  }

  /**
   * Creates a new DB instances and lifts a migration.
   */
  private async createDBInstance(): Promise<DBInstance> {
    try {
      /* Constants */

      const id = Math.random()
        .toString(36)
        .slice(2)
      const tmpDir = os.tmpdir()
      const datasources: DataSource[] = [
        {
          name: 'db',
          connectorType: 'sqlite',
          url: `file:${tmpDir}`,
          config: {},
        },
      ]
      const dbFolder = path.join(tmpDir, './db/')

      /* Occupy resource pool. */
      this.dbs.booting = [...this.dbs.booting, id]

      /* Migrate Datamodel */
      const lift = new LiftEngine({
        projectDir: this.cwd,
      })

      const { datamodel } = await lift.convertDmmfToDml({
        dmmf: JSON.stringify(this.dmmf.datamodel),
        dataSources: datasources,
      })

      /* Release resource pool. */
      this.dbs.booting = this.dbs.booting.filter(dbId => dbId !== id)

      return {
        cwd: tmpDir,
        datamodel: datamodel,
      }
    } catch (err) {
      throw err
    }
  }

  /**
   * Returns a Photon instance from the pool once made available, and cleans
   * and populates the instance if necessary.
   *
   * The returned Photon instance includes the db identifier used for
   * the database releasing.
   *
   * @param opts
   */
  async getDBInstance(): Promise<DBInstance> {
    /**
     * Check whether any of the instances is available and creates a new one
     * if the pool limit is not yet reached, otherwise waits for the available
     * instance.
     */
    if (this.dbs.idle.length > 0) {
      /* If available take a resource and return it right away. */
      const [db, ...remaining] = this.dbs.idle
      this.dbs.idle = remaining
      this.dbs.busy = [...this.dbs.busy, db]

      return db
    } else if (
      this.dbs.idle.length + this.dbs.busy.length + this.dbs.booting.length <
      this.capacity
    ) {
      /* If full capacity is not yet reached create new instance. */
      const db = await this.createDBInstance()
      this.dbs.busy = [...this.dbs.busy, db]

      return db
    } else {
      /* Add to the line. */
      const waiter = new Waiter()
      this.waiters = [...this.waiters, waiter]

      return waiter.wait()
    }
  }

  /**
   * Makes the instance available.
   */
  async releaseDBInstance(db: DBInstance): Promise<void> {
    /**
     * Find the busy instance, remove that instance and give
     * it to the next in line.
     */
    const instance = this.dbs.busy.find(bdb => bdb.cwd === db.cwd)
    this.dbs.busy.filter(bdb => bdb.cwd !== db.cwd)

    /* Add to the next in line or make idle. */
    if (this.waiters.length > 0) {
      const [waiter, ...remainingWaiters] = this.waiters
      this.waiters = remainingWaiters
      waiter.allocate(instance)
    } else {
      this.dbs.idle = [...this.dbs.idle, instance]
    }
  }

  /**
   * Drains the pool by deleting all instances.
   */
  async drain(): Promise<void> {
    // TODO
  }
}

class Waiter {
  private promise: Promise<DBInstance>
  private resolve: (dbi: DBInstance) => void

  constructor() {
    this.promise = new Promise((resolve, reject) => {
      this.resolve = resolve
    })
  }

  wait(): Promise<DBInstance> {
    return this.promise
  }

  allocate(instance: DBInstance): void {
    this.resolve(instance)
  }
}
