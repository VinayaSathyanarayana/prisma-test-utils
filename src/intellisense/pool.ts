import { GeneratorOptions } from '@prisma/generator-helper'

/**
 * Returns a dedicated function name matching the specified datasource.
 *
 * @param dmmf
 */
export function generatePoolType(
  options: GeneratorOptions,
): 'getPostgreSQLPool' | 'getMySQLPool' | 'getSQLitePool' {
  const [datasource] = options.datasources

  /* istanbul ignore if */
  if (!datasource) {
    throw new Error(`No defined datasource!`)
  }

  switch (datasource.connectorType) {
    case 'sqlite': {
      return 'getSQLitePool'
    }
    case 'mysql': {
      return 'getMySQLPool'
    }
    case 'postgresql': {
      return 'getPostgreSQLPool'
    }
    /* istanbul ignore next */
    default: {
      throw new Error(`Unknown datasource: ${datasource.connectorType}`)
    }
  }
}
