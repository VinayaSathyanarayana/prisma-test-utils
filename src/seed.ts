import { DMMF } from '@prisma/dmmf'
import Field from '@prisma/dmmf/dist/Field'
import Mapping from '@prisma/dmmf/dist/Mapping'
import Model from '@prisma/dmmf/dist/Model'
import Chance from 'chance'
import _ from 'lodash'
import { Dictionary } from 'lodash'
// import { readPrismaYml, findDatamodelAndComputeSchema } from './datamodel'
import { Faker, FakerBag, FakerSchema, ID, RelationConstraint } from './types'
import { withDefault } from './utils'

export interface SeedOptions {
  seed?: number
  silent?: boolean
  instances?: number
}

/**
 * Seed the database with mock data.
 *
 * @param fakerSchemaDefinition
 * @param opts
 */
export function seed(
  dmmf: DMMF,
  schemaDef?: Faker | SeedOptions,
  _opts?: SeedOptions,
): object[] | Promise<object[]> {
  /* Argument manipulation */

  const __opts = typeof schemaDef === 'object' ? schemaDef : _opts

  const opts = {
    seed: 42,
    silent: false,
    instances: 5,
    ...__opts,
  }

  /* FakerBag, SchemaDefinition */

  const faker = new Chance(_opts.seed)

  const bag: FakerBag = { faker }
  const fakerSchema = typeof schemaDef === 'function' ? schemaDef(bag) : {}

  /* Prisma Model evaluation */

  // const prisma = readPrismaYml()
  // const dmmf = findDatamodelAndComputeSchema(prisma.configPath, prisma.config)
  const photon: any = {}

  /* Fixture calculations */

  const orders: Order[] = getOrdersFromDMMF(dmmf)
  const steps: Step[] = getStepsFromOrders(orders)
  const tasks: Task[] = getTasksFromSteps(steps)
  const fixtures: Fixture[] = getFixturesFromTasks(fakerSchema, tasks)

  const seeds = seedFixturesToDatabase(photon, fixtures, {
    silent: opts.silent,
  })

  return seeds

  /**
   * The Core logic
   * 1. Create orders from models. Orders tell how many instances of a model
   *  should exist in the end.
   * 2. Create steps from orders. Steps cronologically order the creation of the actual
   *  instances to enable relations.
   * 3. Create Tasks from Steps. Tasks represent single unordered unit derived from Step.
   * 4. Convert tasks to virual data instances, apply mock generation functions to obtain actual data, relations are represented as lists
   *  of strings.
   */

  /**
   * Prisma Faker uses intermediate step between model-faker definition
   * and database seeding. Fixture is a virtual data unit used to describe
   * future data and calculate relations.
   */
  type Order = {
    model: Model
    mapping: Mapping
    amount: number
    relations: Dictionary<Relation>
  }

  type RelationType = '1-to-1' | '1-to-many' | 'many-to-1' | 'many-to-many'

  type Relation = {
    type: RelationType
    relationTo: string
    field: Field
    min: number
    max: number
  }

  /**
   * Note the `relationTo` field; it defines the direction of a relation.
   * This helps with the execution process calculation. If the relation is
   * pointing towards the model, we shouldn't create it since the cyclic complement
   * will implement it.
   */
  type Step = {
    order: number // the creation order of a step, starts with 0
    model: Model
    mapping: Mapping
    amount: number // number of instances created in this step
    relations: Dictionary<Relation>
  }

  type Task = {
    order: number // the creation order of a step, starts with 0
    model: Model
    mapping: Mapping
    relations: Dictionary<Relation>
  }

  type FixtureData = Dictionary<
    ID | string | number | boolean | ID[] | string[] | number[] | boolean[]
  >

  /**
   * Represents the virtual unit.
   */
  type Fixture = {
    order: number // starts with 0
    id: string
    model: Model
    mapping: Mapping
    data: FixtureData
    relations: Dictionary<{
      type: RelationType
      relationTo: string // determines the direction of relation
    }>
  }

  /* Helper functions */

  /**
   * Converts dmmf to fixtures. Fixtures represent the summary of a mock requirements
   * for particular model.
   *
   * @param dmmf
   */
  function getOrdersFromDMMF(dmmf: DMMF): Order[] {
    return dmmf.datamodel.models.map(model => {
      /* User defined settings */
      const fakerModel = withDefault(
        {
          amount: opts.instances,
          factory: undefined,
        },
        fakerSchema[model.name],
      )

      /* Find Photon mappings for seeding step */
      const mapping = dmmf.mappings.find(m => m.model === model.name)

      /* Generate relations based on provided restrictions. */
      const relations: Order['relations'] = model.fields
        .filter(f => f.isRelation())
        .reduce<Order['relations']>((acc, field) => {
          const fakerField: RelationConstraint = _.get(
            fakerModel,
            ['factory', field.name],
            {
              min: 1,
              max: 1,
            },
          )

          switch (typeof fakerField) {
            case 'object': {
              /* Calculate the relation properties */
              const { type, min, max } = getRelationType(
                dmmf.datamodel.models,
                field,
                fakerField,
              )

              return {
                ...acc,
                [field.name]: {
                  type: type,
                  min: min,
                  max: max,
                  relationTo: '',
                  field,
                },
              }
            }
            default: {
              throw new Error(
                `Expected a relation constraint got ${typeof fakerField}`,
              )
            }
          }
        }, {})

      return {
        model: model,
        mapping: mapping,
        amount: fakerModel.amount,
        relations: relations,
      }
    })

    /**
     * Derives the relation type and constraints from the field
     * and definition.
     *
     * We have four different relation types:
     * 1. 1-to-1
     *  ~ There can be at most one connection or none if optional.
     *  ~ We'll connect the nodes from this model (a child model).
     * 2. 1-to-many
     *  ~ The node can have 0 to infinite connections (we'll use min/max spec).
     *  ~ We'll connect the nodes from this model (a child model).
     * 3. many-to-1
     *  ~ A particular instance can either connect or not connect if optional.
     *  ~ We'll fill the pool with ids of this model (a parent model).
     * 4. many-to-many
     *  ~ A particular instance can have 0 to infinite connections (we'll use min/max spec).
     *  ~ We'll use `relationTo` to determine whether this is a parent or child model.
     *
     * We presume that the field is a relation.
     */
    function getRelationType(
      allModels: Model[],
      field: Field,
      definition: RelationConstraint,
    ): {
      type: RelationType
      min: number
      max: number
    } {
      /* Models */

      const fieldModel = field.getModel()
      const relationModel = allModels.find(m => m.name === field.type)

      const relationField = withDefault<Field>(
        {
          kind: '',
          name: '',
          isRequired: true,
          isList: false,
          isId: false,
          type: field.type,
        } as Field,
        relationModel.fields.find(f => f.type === fieldModel.name),
      )

      if (field.isList && relationField.isList) {
        /**
         * many-to-many (A)
         *
         * model A {
         *  bs: [B]
         * }
         * model B {
         *  as: [A]
         * }
         */
        const min = withDefault(0, definition.min)
        const max = withDefault(min, definition.max)

        return {
          type: 'many-to-many',
          min: min,
          max: max,
        }
      } else if (!field.isList && relationField.isList) {
        /**
         * many-to-1 (A)
         *
         * model A {
         *  b: B
         * }
         * model B {
         *  as: [A]
         * }
         */

        return {
          type: 'many-to-1',
          min: field.isRequired ? 1 : 0,
          max: 1,
        }
      } else if (field.isList && !relationField.isList) {
        /**
         * 1-to-many (A)
         *
         * model A {
         *  bs: [B]
         * }
         *
         * model B {
         *  a: A
         * }
         */

        const min = withDefault(field.isRequired ? 1 : 0, definition.min)
        const max = withDefault(min, definition.max)

        return {
          type: '1-to-many',
          min: min,
          max: max,
        }
      } else {
        /**
         * 1-to-1 (A)
         *
         * model A {
         *  b: B
         * }
         * model B {
         *  a: A
         * }
         */

        return {
          type: '1-to-1',
          min: field.isRequired ? 1 : 0,
          max: 1,
        }
      }
    }
  }

  /**
   * Coverts fixtures to steps. Steps represent an ordered entity that specifies the creation
   * of a virtual data.
   *
   * Creates a pool of available instances and validates relation constraints.
   *
   * @param fixtures
   */
  function getStepsFromOrders(orders: Order[]): Step[] {
    type Pool = Dictionary<{
      model: Model
      units: number
    }>

    /* Triggers the order to steps conversion */

    const steps = sort(orders, orders, [], {})

    return steps

    /* Helper functions */

    /**
     * The sort function functionally implements topological sort algorithm
     * by making sure all relations have been defined prior to the inclusion of
     * an order in the chain.
     */
    function sort(
      allOrders: Order[],
      remainingOrders: Order[],
      sortedSteps: Step[],
      pool: Pool,
    ): Step[] {
      switch (remainingOrders.length) {
        case 0: {
          return []
        }
        case 1: {
          const [o] = remainingOrders

          /* Checks if the order is well defined */
          if (isOrderWellDefinedInPool(pool, o)) {
            const [steps] = getStepsFromOrder(allOrders, sortedSteps, pool, o)

            return steps
          } else {
            /** Since this is the last order, we cannot obtain any new resources,
             * meaning that there's an issue with the order.
             */
            throw new Error(`${o.model.name} uses undefined relations!`)
          }
        }
        default: {
          const [o, ...os] = remainingOrders

          /* Checks if the order is well defined */
          if (isOrderWellDefinedInPool(pool, o)) {
            const [steps, newPool] = getStepsFromOrder(
              allOrders,
              sortedSteps,
              pool,
              o,
            )

            return [
              ...steps,
              ...sort(allOrders, os, [...sortedSteps, ...steps], newPool),
            ]
          } else {
            /**
             * If the order is not yet well defined, we put it at the end of the list
             * and wait for resources in the pool to be adequate.
             */
            return sort(allOrders, [...os, o], sortedSteps, pool)
          }
        }
      }
    }

    /**
     * Determines whether we can already process the order based on the pool
     * capacity.
     *
     * This function in combination with `getStepsFromOrder` should give you
     * all you need to implement meaningful topological sort on steps.
     */
    function isOrderWellDefinedInPool(pool: Pool, order: Order) {
      return Object.values(order.relations).every(
        relation =>
          pool.hasOwnProperty(relation.field.type) ||
          relation.type === 'many-to-1' ||
          relation.type === 'many-to-many',
      )
    }

    /**
     * Converts a well defined order to multiple steps.
     *
     * This function in combination with `isOrderWellDefinedInPool` should give
     * you everything you need to implement meaningful topological sort on steps.
     */
    function getStepsFromOrder(
      allOrders: Order[],
      sortedSteps: Step[],
      pool: Pool,
      order: Order,
    ): [Step[], Pool] {
      /**
       * A step unit derived from the order.
       */
      const step: Step = {
        order: sortedSteps.length,
        amount: order.amount,
        model: order.model,
        mapping: order.mapping,
        relations: order.relations,
      }

      /**
       * Assumes that there cannot exist two models with the same name.
       */
      const newPool: Pool = {
        ...pool,
        [order.model.name]: {
          model: order.model,
          units: order.amount,
        },
      }

      return [[step], newPool]
    }
  }

  /**
   * Converts steps to tasks.
   *
   * Steps to Tasks introduce no non-trivial logic. It's a simple conversion mechanism
   * to make system more robuts and make the mocking part more granular.
   *
   * @param steps
   */
  function getTasksFromSteps(steps: Step[]): Task[] {
    const tasks = steps.reduce<Task[]>((acc, step) => {
      const intermediateTasks: Task[] = Array<Task>(step.amount).fill({
        order: step.order,
        model: step.model,
        mapping: step.mapping,
        relations: step.relations,
      })

      return acc.concat(...intermediateTasks)
    }, [])

    return tasks
  }

  /**
   * Converts fixtures from steps by creating a pool of available instances
   * and assigning relations to particular types.
   *
   * This function assumes that:
   *  1. Tasks (Steps) are sorted in such an order that pool always possesses
   *    all required instances,
   *  2. There are enough resources in the pool at all times,
   *  3. The provided schema is valid.
   *
   * @param steps
   */
  function getFixturesFromTasks(schema: FakerSchema, tasks: Task[]): Fixture[] {
    /**
     * Pool describes the resources made available by a parent type to its children.
     */
    type Pool = Dictionary<{ [child: string]: ID[] }>

    const [fixtures] = _.sortBy(tasks, t => t.order).reduce<[Fixture[], Pool]>(
      ([fixtures, pool], task) => {
        const id = faker.guid()
        const [data, newPool] = getMockDataForTask(id, pool, task)

        const fixture: Fixture = {
          order: fixtures.length,
          id: id,
          model: task.model,
          mapping: task.mapping,
          data: data,
          relations: task.relations,
        }

        return [fixtures.concat(fixture), newPool]
      },
      [[], {}],
    )

    return fixtures

    /* Helper functions */

    /**
     * Generates mock data from the provided model. Scalars return a mock scalar or
     * list of mock scalars, relations return an ID or lists of IDs.
     */
    function getMockDataForTask(
      id: ID,
      _pool: Pool,
      task: Task,
    ): [FixtureData, Pool] {
      const [finalPool, fixture] = task.model.fields.reduce(
        ([pool, acc], field) => {
          const fieldModel = field.getModel()

          /* Custom field mocks */

          if (
            schema[task.model.name] &&
            schema[task.model.name].factory &&
            schema[task.model.name].factory[field.name]
          ) {
            const mock = schema[task.model.name].factory[field.name]
            switch (typeof mock) {
              case 'function': {
                const value = mock()
                return [pool, { ...acc, [field.name]: value }]
              }
              case 'object': {
                /* Relation constraint */
                break
              }
              default: {
                const value = mock
                return [pool, { ...acc, [field.name]: value }]
              }
            }
          }

          switch (field.type) {
            case 'ID': {
              const id = faker.guid()

              return [pool, { ...acc, [field.name]: id }]
            }
            case 'String': {
              const string = faker.word()

              return [pool, { ...acc, [field.name]: string }]
            }
            case 'Int': {
              const number = faker.integer()

              return [pool, { ...acc, [field.name]: number }]
            }
            case 'Float': {
              const float = faker.floating()

              return [pool, { ...acc, [field.name]: float }]
            }
            case 'Date': {
              const date = faker.date()

              return [pool, { ...acc, [field.name]: date }]
            }
            default: {
              /* Relations */
              if (field.isRelation()) {
                /* Resources calculation */
                const relation = task.relations[field.name]
                const units = faker.integer({
                  min: relation.min,
                  max: relation.max,
                })

                switch (relation.type) {
                  case '1-to-1': {
                    /**
                     * 1-to-1 relation should take at most one id from the resource pool
                     * and submit no new ids. Because we already manage requirements during
                     * order creation step, we can ignore it now.
                     */
                    const [newPool, ids] = getIDInstancesFromPool(
                      pool,
                      fieldModel.name,
                      field.type,
                      units,
                    )

                    /**
                     * This makes sure that relations are properly connected.
                     */
                    switch (ids.length) {
                      case 0: {
                        return [newPool, acc]
                      }

                      case 1: {
                        const [id] = ids
                        return [
                          newPool,
                          { ...acc, [field.name]: { connect: { id } } },
                        ]
                      }

                      default: {
                        throw new Error(`Something truly unexpected happened.`)
                      }
                    }
                  }
                  case '1-to-many': {
                    /**
                     * 1-to-many takes from 0 or 1 to min/max number of ids from the pool
                     * and creates no new ids along the way. We don't have to worry about that
                     * though since we have taken care of everything during the order generation
                     * step.
                     */
                    const [newPool, ids] = getIDInstancesFromPool(
                      pool,
                      fieldModel.name,
                      field.type,
                      units,
                    )

                    const connections = ids.reduce((acc, id) => {
                      return [...acc, { id }]
                    }, [])

                    return [
                      newPool,
                      {
                        ...acc,
                        [field.name]: {
                          connect: connections,
                        },
                      },
                    ]
                  }
                  case 'many-to-1': {
                    /**
                     * Many-to-1 relations only give out ids to be later
                     * accessed by 1-to relations.
                     */

                    const newPool = insertIDInstancesIntoPool(
                      pool,
                      field.type,
                      fieldModel.name,
                      id,
                      units,
                    )

                    return [newPool, acc]
                  }
                  case 'many-to-many': {
                    // TODO: Implement `relationTo`!
                    return [pool, acc]
                  }
                }
              }

              /* Fallback for unsupported scalars */
              throw new Error(`Unsupported field type "${field.type}".`)
            }
          }
        },
        [_pool, {}],
      )

      return [fixture, finalPool]
    }

    // TODO: Is it possible that a relation gets multiple same ids?

    /**
     * Retrieves an ID from the pool and removes its instance.
     */
    function getIDInstancesFromPool(
      pool: Pool,
      parent: string,
      child: string,
      n: number = 1,
    ): [Pool, ID[]] {
      const ids = _.get(pool, [parent, child])
      return [_.set(pool, [parent, child], ids.splice(n)), ids.slice(0, n)]
    }

    /**
     * Inserts n-replications of ID into the pool and returns the new pool.
     */
    function insertIDInstancesIntoPool(
      pool: Pool,
      parent: string,
      child: string,
      id: ID,
      n: number = 1,
    ): Pool {
      const ids = _.get(pool, [parent, child], [])
      return _.set(pool, [parent, child], [...ids, ...Array(n).fill(id)])
    }
  }

  /**
   * Seeds the fixtures to the database. Based on the `silent` option
   * it performs data push. Photon is provided globally.
   *
   * @param fixtures
   * @param opts
   */
  function seedFixturesToDatabase(
    photon: any,
    fixtures: Fixture[],
    opts: { silent: boolean } = { silent: false },
  ): object[] | Promise<object[]> {
    if (opts.silent) {
      return _.sortBy(fixtures, f => f.order).map(f => f.data)
    } else {
      /**
       * Generates a chain of promises that create DB instances.
       */
      const actions = _.sortBy(fixtures, f => f.order).reduce<
        Promise<object[]>
      >(async (acc, f) => {
        return acc.then(async res => {
          const seed = await photon[f.mapping.create]({ data: f.data })
          return res.concat(seed)
        })
      }, Promise.resolve([]))
      return actions
    }
  }
}
