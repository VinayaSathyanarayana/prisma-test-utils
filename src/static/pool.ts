import { LiftEngine, DataSource } from '@prisma/lift'
import { DMMF } from '@prisma/photon/runtime/dmmf-types'
import * as os from 'os'
import * as path from 'path'
import { withDefault } from '../utils'

export type PoolOptions = {
  pool: {
    max?: number
  }
}

export type DBInstance = {
  url: string
  cwd: string
  datamodel: string
}

export default (dmmf: DMMF.Document) =>
  class Pool {
    private dbs: {
      booting: string[]
      idle: DBInstance[]
      busy: DBInstance[]
    }
    private waiters: Waiter[]
    private capacity: number

    constructor(definition: PoolOptions) {
      this.dbs = {
        booting: [],
        idle: [],
        busy: [],
      }
      this.waiters = []
      this.capacity = withDefault(Infinity, definition.pool.max)
    }

    /**
     * Creates a new DB instances and lifts a migration.
     */
    private async createDBInstance(): Promise<void> {
      try {
        /* Constants */

        const id = Math.random()
          .toString(36)
          .slice(2)
        const tmpDir = path.join(os.tmpdir(), `prisma-pool-${id}`)
        const dbFile = path.join(tmpDir, './db.db')
        const datasources: DataSource[] = [
          {
            name: 'db',
            connectorType: 'sqlite',
            url: `file:${dbFile}`,
            config: {},
          },
        ]

        /* Occupy resource pool. */
        this.dbs.booting = [...this.dbs.booting, id]

        /* Migrate Datamodel. */
        const lift = new LiftEngine({
          projectDir: tmpDir,
        })

        const datamodelDmmf = {
          enums: [],
          models: [],
          ...dmmf.datamodel,
        }

        const { datamodel } = await lift.convertDmmfToDml({
          dmmf: JSON.stringify(datamodelDmmf),
          config: { datasources, generators: [] },
        })

        const {
          datamodelSteps,
          errors: stepErrors,
        } = await lift.inferMigrationSteps({
          migrationId: id,
          datamodel: datamodel,
          assumeToBeApplied: [],
          sourceConfig: datamodel,
        })

        if (stepErrors.length > 0) {
          throw stepErrors
        }

        const { errors } = await lift.applyMigration({
          force: true,
          migrationId: id,
          steps: datamodelSteps,
          sourceConfig: datamodel,
        })

        if (errors.length > 0) {
          throw errors
        }

        const progress = () =>
          lift.migrationProgess({
            migrationId: id,
            sourceConfig: datamodel,
          })

        while ((await progress()).status !== 'MigrationSuccess') {
          /* Just wait */
        }

        /* Release resource pool. */
        this.dbs.booting = this.dbs.booting.filter(dbId => dbId !== id)

        const instance: DBInstance = {
          url: dbFile,
          cwd: tmpDir,
          datamodel: datamodel,
        }

        /**
         * If there's a waiter in line allocate an instance, otherwise make it idle.
         */
        if (this.waiters.length > 0) {
          const [waiter, ...remainingWaiters] = this.waiters
          this.waiters = remainingWaiters
          this.dbs.busy = [...this.dbs.busy, instance]
          waiter.allocate(instance)
        } else {
          this.dbs.idle = [...this.dbs.idle, instance]
        }
      } catch (err) {
        throw err
      }
    }

    /**
     * Run the encapsulated funciton.
     *
     * @param fn
     */
    async run<T>(fn: (db: DBInstance) => Promise<T>) {
      const db = await this.getDBInstance()
      try {
        const result = await fn(db)
        await this.releaseDBInstance(db)
        return result
      } catch (e) {
        await this.releaseDBInstance(db)
        throw e
      }
    }

    /**
     * Returns a Photon instance from the pool once made available, and cleans
     * and populates the instance if necessary.
     *
     * The returned Photon instance includes the db identifier used for
     * the database releasing.
     *
     * @param opts
     */
    async getDBInstance(): Promise<DBInstance> {
      /**
       * Check whether any of the instances is available and creates a new one
       * if the pool limit is not yet reached, otherwise waits for the available
       * instance.
       */
      if (this.dbs.idle.length > 0) {
        /* If available take a resource and return it right away. */
        const [db, ...remaining] = this.dbs.idle
        this.dbs.idle = remaining
        this.dbs.busy = [...this.dbs.busy, db]

        return db
      } else {
        /* Add to the line. */
        const waiter = new Waiter()
        this.waiters = [...this.waiters, waiter]

        /**
         * If full capacity is not yet reached create new instance.
         * Otherwise, throw an error.
         */
        if (
          this.dbs.idle.length +
            this.dbs.busy.length +
            this.dbs.booting.length <
          this.capacity
        ) {
          this.createDBInstance()
        } else {
          throw new Error(`You've reached the upper limit of instances.`)
        }

        return waiter.wait()
      }
    }

    /**
     * Makes the instance available.
     */
    async releaseDBInstance(db: DBInstance): Promise<void> {
      /**
       * Find the busy instance, remove that instance and give
       * it to the next in line.
       */
      const instance = this.dbs.busy.find(bdb => bdb.cwd === db.cwd)!
      this.dbs.busy = this.dbs.busy.filter(bdb => bdb.cwd !== db.cwd)

      /* Allocate to the next waiter in line or make idle. */
      // if (this.waiters.length > 0) {
      //   const [waiter, ...remainingWaiters] = this.waiters
      //   this.waiters = remainingWaiters
      //   waiter.allocate(instance)
      // } else {
      //   this.dbs.idle = [...this.dbs.idle, instance]
      // }

      this.dbs.idle = [...this.dbs.idle, instance]
    }

    /**
     * Drains the pool by deleting all instances.
     */
    async drain(): Promise<void> {
      // TODO
    }
  }

class Waiter {
  private promise: Promise<DBInstance>
  private resolve: (dbi: DBInstance) => void = () => {}

  constructor() {
    this.promise = new Promise((resolve, reject) => {
      this.resolve = resolve
    })
  }

  wait(): Promise<DBInstance> {
    return this.promise
  }

  allocate(instance: DBInstance): void {
    this.resolve(instance)
  }
}
