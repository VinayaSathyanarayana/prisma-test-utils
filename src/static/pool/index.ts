export { getPostgreSQLPool, PostgreSQLPoolOptions } from './dbs/postgres'
export { getMySQLPool, MySQLPoolOptions } from './dbs/mysql'
export { getSQLitePool, SQLitePoolOptions, getTmpSQLiteDB } from './dbs/sqlite'
