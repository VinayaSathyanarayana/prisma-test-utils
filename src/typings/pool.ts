import { GeneratorOptions } from '@prisma/cli'

/**
 * Returns a dedicated function name matching the specified datasource.
 *
 * @param dmmf
 */
export function generatePoolType(
  options: GeneratorOptions,
): 'getPostgreSQLPool' | 'getMySQLPool' | 'getSQLitePool' {
  const datasource = options.dataSources

  console.log(datasource)
  return 'getMySQLPool'
}
