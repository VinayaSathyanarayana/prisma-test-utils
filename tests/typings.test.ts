import { getDMMF, getConfig } from '@prisma/photon'
import { DMMF } from '@prisma/photon/runtime/dmmf-types'

import { generatePoolType } from '../src/typings/pool'
import { generateGeneratedSeedModelsType } from '../src/typings/seed'

describe('typings:', () => {
  test('generates seed types correctly', async () => {
    const datamodel = `
datasource sqlite {
  url      = "file:./data.db"
  provider = "sqlite"
}

generator photonjs {
  provider = "photonjs"
  output   = "./@generated/photon"
}

generator testutils {
  provider = "prisma-test-utils"
  output   = "./@generated/prisma-test-utils"
}

model User {
  id       String  @id @default(cuid())
  name     String
  email    String
  isActive Boolean
  pet      Pet
  house    House?
  friends  User[]
}

enum Animal {
  Dog
  Cat
  Dinosaur
}

model Pet {
  id     String @id @default(cuid())
  name   String
  animal Animal
  toys   Toy[]
}

model Toy {
  id    String @id @default(cuid())
  name  String
  price Float
}

model House {
  id            String @id @default(cuid())
  address       String
  numberOfRooms Int
  residents     User[]
}
  `
    const dmmf = await getDMMF({ datamodel })
    expect(generateGeneratedSeedModelsType(dmmf)).toMatchSnapshot()
  })

  test('generates correct sqlite pool type', async () => {
    const datamodel = `
datasource sqlite {
  url      = "file:./data.db"
  provider = "sqlite"
}

model User {
  id       String  @id @default(cuid())
  name     String
}
  `
    const options = await getConfig(datamodel)
    expect(
      generatePoolType({
        cwd: '',
        datamodel,
        dataSources: options.datasources,
        dmmf: null,
        generator: {
          output: '',
          name: '',
          provider: '',
          platforms: [],
          config: {},
        },
        otherGenerators: [],
      }),
    ).toMatchSnapshot()
  })

  test('generates correct mysql pool type', async () => {
    const datamodel = `
datasource mysql {
  url      = "mysql://user@localhost:3333"
  provider = "mysql"
}

model User {
  id       String  @id @default(cuid())
  name     String
}
  `
    const options = await getConfig(datamodel)
    expect(
      generatePoolType({
        cwd: '',
        datamodel,
        dataSources: options.datasources,
        dmmf: null,
        generator: {
          output: '',
          name: '',
          provider: '',
          platforms: [],
          config: {},
        },
        otherGenerators: [],
      }),
    ).toMatchSnapshot()
  })

  test('generates correct postgresql pool type', async () => {
    const datamodel = `
datasource postgresql {
  url      = "postgresql://user:secret@localhost"
  provider = "postgresql"
}

model User {
  id       String  @id @default(cuid())
  name     String
}
  `
    const options = await getConfig(datamodel)
    expect(
      generatePoolType({
        cwd: '',
        datamodel,
        dataSources: options.datasources,
        dmmf: null,
        generator: {
          output: '',
          name: '',
          provider: '',
          platforms: [],
          config: {},
        },
        otherGenerators: [],
      }),
    ).toMatchSnapshot()
  })
})
