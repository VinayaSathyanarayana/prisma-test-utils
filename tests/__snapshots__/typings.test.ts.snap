// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typings: correctly errors on missing datasources 1`] = `"No defined datasource!"`;

exports[`typings: generates correct mysql pool type 1`] = `"getMySQLPool"`;

exports[`typings: generates correct postgresql pool type 1`] = `"getPostgreSQLPool"`;

exports[`typings: generates correct sqlite pool type 1`] = `"getSQLitePool"`;

exports[`typings: generates seed types correctly 1`] = `
"\\"*\\": { amount: number }
User: { 
  amount?: number, 
  factory?: {
    id?: () => string
name?: () => string
email?: () => string
isActive?: () => boolean
friends?: { min?: number, max?: number } 
  }
}
Pet: { 
  amount?: number, 
  factory: {
    id?: () => string
name?: () => string
animal: () => \\"Dog\\" | \\"Cat\\" | \\"Dinosaur\\"
toys?: { min?: number, max?: number } 
  }
}
Toy: { 
  amount?: number, 
  factory?: {
    id?: () => string
name?: () => string
price?: () => number 
  }
}
House: { 
  amount?: number, 
  factory?: {
    id?: () => string
address?: () => string
numberOfRooms?: () => number
residents?: { min?: number, max?: number } 
  }
}"
`;
